src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,464 @@
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_freq_status.at(domain_idx).value = m_platform_io.sample(m_gpu_freq_status.at(
+                                                                          domain_idx).batch_idx);
+            m_gpu_compute_activity.at(domain_idx).value = m_platform_io.sample(m_gpu_compute_activity.at(
+                                                                               domain_idx).batch_idx);
+            m_gpu_utilization.at(domain_idx).value = m_platform_io.sample(m_gpu_utilization.at(
+                                                                          domain_idx).batch_idx);
+
+            m_gpu_energy.at(domain_idx).value = m_platform_io.sample(m_gpu_energy.at(
+                                                                     domain_idx).batch_idx);
+        }
+
+        m_time.value = m_platform_io.sample(m_time.batch_idx);
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence
+    void GPUActivityAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
+        while(geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
+        geopm_time(&m_last_wait);
+    }
+
+    // Adds the wait time to the top of the report
+    std::vector<std::pair<std::string, std::string> > GPUActivityAgent::report_header(void) const
+    {
+        return {{"Wait time (sec)", std::to_string(M_WAIT_SEC)}};
+    }
+
+    // Adds number of frquency requests to the per-node section of the report
+    std::vector<std::pair<std::string, std::string> > GPUActivityAgent::report_host(void) const
+    {
+        std::vector<std::pair<std::string, std::string> > result;
+
+        result.push_back({"GPU Frequency Requests", std::to_string(m_gpu_frequency_requests)});
+#ifdef GEOPM_DEBUG
+        result.push_back({"GPU Clipped Frequency Requests", std::to_string(m_gpu_frequency_clipped)});
+#endif
+        result.push_back({"Resolved Max Frequency", std::to_string(m_f_max)});
+        result.push_back({"Resolved Efficient Frequency", std::to_string(m_f_efficient)});
+        result.push_back({"Resolved Frequency Range", std::to_string(m_f_range)});
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            double energy_stop = m_gpu_active_energy_stop.at(domain_idx);
+            double energy_start = m_gpu_active_energy_start.at(domain_idx);
+            double region_stop = m_gpu_active_region_stop.at(domain_idx);
+            double region_start =  m_gpu_active_region_start.at(domain_idx);
+            result.push_back({"GPU " + std::to_string(domain_idx) +
+                              " Active Region Energy", std::to_string(energy_stop - energy_start)});
+            result.push_back({"GPU " + std::to_string(domain_idx) +
+                              " Active Region Time", std::to_string(region_stop - region_start)});
+#ifdef GEOPM_DEBUG
```

Comment:
Again, these region times might look pretty different between a debug and release build.  If that matters, I'd just do this all the time.

Reply:
These I'd omit, as ACTIVE REGION TIME provides most of the relevant information.